src/Swarm/Util/WindowedCounter.hs

```diff
@@ -0,0 +1,135 @@
+-- that is still stored falls outside of the nominal retention window.
+--
+-- One might think we could perform garbage collection whenever we execute queries.
+-- However, in the context in which the view powered by the query is generated, we
+-- are not permitted to mutate the "state" of the game
+-- (the type signature of the rendering function is @AppState -> Widget Name@).
+--
+-- Therefore, when we perform "queries" on the window, we must apply some
+-- filtering to exclude the "stragglers"; data members that have already fallen outside
+-- the window but have not yet been "garbage collected".
+-- We use a 'Set' to allow this filtering to be performed in @O(log n)@ time.
+--
+-- In the worst case, the entire dataset may "age out" without being garbage collected,
+-- so that an @O(log n)@ filtering operation might be performed upon every "frame refresh"
+-- of the UI view.
+-- However, we also store the largest element of the window separately from the 'Set' so that
+-- we can compare against it for a @O(1)@ short-circuited path once every member ages out.
+data WindowedCounter a = WindowedCounter
+  { _members :: Set a
+  , _lastLargest :: Maybe a
+  -- ^ NOTE: It is possible that '_lastLargest' may not exist in the 'Set'.
+  , _nominalSpan :: a
+  -- ^ Data retention window
+  }
+  -- NOTE: deriving 'FromJSON' circumvents the protection offered by "smart constructors",
+  -- and the 'ToJSON' instance may expose internal details.
+  deriving (Eq, Show, Generic, FromJSON, ToJSON)
+
+mkWindow ::
+  -- | Window span
+  a ->
+  WindowedCounter a
+mkWindow = WindowedCounter Set.empty Nothing
+
+-- | Return the ratio of {members in the window} to the {integral span
+-- represented by the window} (actually, the beginning of the window up to
+-- some reference tick).
+--
+-- The "reference tick" must be at least as large as the largest
+-- element of the window.
+--
+-- A fully-contiguous collection of ticks would have an occupancy ratio of @1@.
+getOccupancy ::
+  Integral a =>
+  -- | reference tick
+  a ->
+  WindowedCounter a ->
+  Ratio a
+getOccupancy referenceTick (WindowedCounter s lastLargest nominalSpan) =
+  if Set.null s || maybe False (< referenceTick - nominalSpan) lastLargest
+    then 0
+    else case Set.lookupMin s of
+      Nothing -> 0
+      Just minElement -> fromIntegral (Set.size s) % (referenceTick - minElement)
+
+-- | Invocations of this function shall be guarded externally
+-- by the conditions meant to be tracked in the window.
+--
+-- The value inserted must always be at least as large as the
+-- current largest element of the set!
+-- If it is equal, it is ignored.
+--
+-- The 'discardGarbage' function is called from inside this function
+-- so that maintenance of the data structure is simplified.
+insert :: (Show a, Integral a) => a -> WindowedCounter a -> WindowedCounter a
+insert x statusQuo@(WindowedCounter s lastLargest nominalSpan)
+  | maybe False (x <) lastLargest =
+      error $
```

Comment:
I really, really don't want to have calls to `error` in the codebase, because it means that if there is a bug it could end up crashing the entire game, so someone could lose all their progress.  In this particular case, some alternatives might include (1) just ignore invalid inserts; (2) return a `Maybe (WindowedCounter a)` and check for the error condition elsewhere; (3) create some kind of "strict checking mode", enabled by a flag, where we do crash when encountering such a bug, and we can turn on that flag in our CI tests, but have it off by default (in which case we simply ignore invalid inserts as in option (1)).

Reply:
Option (3) sounds like #217?